### DISCUSSION TITLE
Only left side of the keyboard works

### DISCUSSION THREAD

**Discussion by hasufell** [dpapavas/lagrange-keyboard#9] [Q&A] [answered]
## Problem

I'm done wiring the matrix and assembling. I successfully flashed the board and the left keyboard seems to be working. The right board is detected, can be flashed etc, but doesn't react to key presses. I checked that the pins on the molex receive the keypresses and they do (rows and columns). I'm suspecting this just leaves the option that something is wrong with the controller... but if so, why would it be detected and flashable?

## Pictures
<details>
  <summary>Click to expand!</summary>
  
![IMG_20210505_120219](https://user-images.githubusercontent.com/1241845/117125762-3f176a00-ad9a-11eb-9020-5300b1913fc9.jpg)
![IMG_20210505_120239](https://user-images.githubusercontent.com/1241845/117125783-476fa500-ad9a-11eb-9463-01fb2ef299ed.jpg)
![IMG_20210505_120300](https://user-images.githubusercontent.com/1241845/117125847-59e9de80-ad9a-11eb-8f50-2dbc4486d3d0.jpg)

</details>

**Comment by dpapavas** [comment]
Most likely, something is wrong with the driver!  A change made in the split transport layer of QMK after my initial driver submission (coupled with an oversight on my part, that allowed it to happen unnoticed) led to a problem with the effects you describe (although I would have expected both sides to work when they are connected via USB, i.e. the master should work regardless of which one it is).  See [here](https://github.com/qmk/qmk_firmware/pull/12606#issuecomment-823459873) for more details.

If you don't have the latest QMK version, please upgrade to it and try again.  I just flashed version `0.12.43` of the QMK master branch on my keyboard and am typing this on it with no problems.

Also: congratulations on your build; it looks very nicely made!  Did you have any particular difficulty with any of the wiring steps?  How long did it take you to finish each side?

**Comment by hasufell** [comment]
Ok, so I flashed 0.12.43 from main qmk repository. Still no change.

What I'm observing is this:

1. USB plugged into left board: left board works as expected, right board doesn't react to key presses
2. USB plugged into right board: left board seems to be a mirrored version of what should be the right board (so `T` key for example produces `Y`), right board still doesn't react to key presses

The way I flashed was this:
1. press reset button on the master
2. run `make handwired/lagrange:default:dfu`

I also re-ran `make git-submodules` to be sure.

**Comment by dpapavas** [comment, marked as answer]
What you report seems to suggest that everything is wired and working properly on the left side and that at least the controller is working on the right side and communicating properly both with the PC and with the other half.

I don't see how this could be to sole source of your problem, but I notice that I've ommited a step in the instructions in the build guide:  using the default configuration, which uses the `EE_HANDS` method (see `config.h`) to set handedness, you need to flash the EEPROM on each side to mark which is which.  See [here](https://docs.qmk.fm/#/feature_split_keyboard?id=handedness-by-eeprom) for more information.  It should be enough to flash the left side with `make handwired/lagrange:default:dfu-split-left` and the right side with `make handwired/lagrange:default:dfu-split-right`.  This should at least result in the left side consistently registering `T`, regardless of which side is plugged in, but you best try it out before we investigate further, since it might conceivably help with your problem as well.

**Reply by hasufell** [reply]
>  It should be enough to flash the left side with make handwired/lagrange:default:dfu-split-left and the right side with make handwired/lagrange:default:dfu-split-right. This should at least result in the left side consistently registering T, regardless of which side is plugged in, but you best try it out before we investigate further, since it might conceivably help with your problem as well.

Excellent... that did it indeed! I'm surprised I didn't have to re-solder much. Desoldering is one of my worst nightmares.

> Also: congratulations on your build; it looks very nicely made! Did you have any particular difficulty with any of the wiring steps? How long did it take you to finish each side?

I think there are 3 main difficulties:

1. **Forming the loops** that you push onto the pins later... my switches had two kinds of pins, one very stiff and one quite bendable, which also isn't really a circle if you would cut it horizontally. The bendable pin was pretty hard to deal with and you only have 4 to 5 tries before the pin just breaks off completely. I'm not sure if all switches are like that. In the beginning, I used a free switch to form the loops (as suggesged in the build guide), later (especially when wiring the rows) I just used precision tweezer and generally ended up with slightly too big loops, which then required a lot of solder.
2. **Soldering the multi-stranded wire** is really painful and was the only place where I had to re-solder quite a few, because it wasn't connected properly. The wire overheats quickly and may turn into black dust. I tried the technique of putting a drop of solder onto the iron and just dropping it between the multi-stranded wire and the solid-core wire, but that also only worked a couple of times. I guess you want to heat the solid-core wire primarily, because it's less temperature prone, but that's also hard when you're trying to wrap the multi-stranded one around it. My end result doesn't look particularly pretty, but my plan is to keep the keyboard case closed, so no one will notice :sweat_smile: 
3. **Crimping** was pretty rough. I felt like I got the hang of it really well though, since I've dealt with re-crimping stuff in the internals of my soldering station, which had a broken on/off switch. However, the really hard part is making it stick well on the molex. When connecting it to the board, some wires (along with the crimp terminal) would get pushed out again. I'm assuming because the exposed part of the wire was pushed too far into the crimp terminal, blocking the small hole for the pin. However, I was expecting the crimp terminals to be much better attached to the molex and not come out from that little force. Maybe I missed something there.

Oh, and  another thing that was hard was dealing with the insulation for the solid core wire when doing the rows. I alternated between just pushing the insulation further after cutting it (which really requires a lot of force and careful handling... sometimes I would push out the whole insulation by accident and then have to start over) and pushing a sharp tweezer needle into the insulation, which then allowed me to just rip it off downwards. That kinda proved to be the better approach, except that using a lot of force with a sharp tweezer can potentially deform it and is also pretty dangerous. I rammed it at least once into one of my finger nails... and it pierced right through it :smile:

----

The only thing that's left is printing the special keys. I haven't had much luck with that so far, because they don't fit onto my switches and I didn't have small enough file to clean that up. I considered buying a resin printer, but I'm not sure I want to get into another episode of experimentation right now.

One thing I wanted to try and will probably open another issue about it is that the fitting could be improved by using a different bottom piece, such as the one from the KeyV2 project: https://github.com/rsheldiii/KeyV2

![Screenshot_2021-05-05_14-31-03](https://user-images.githubusercontent.com/1241845/117141260-9e7f7500-adae-11eb-824d-4551b7e08173.png)

**Comment by dpapavas** [comment]
> In the beginning, I used a free switch to form the loops (as suggested in the build guide), later (especially when wiring the rows) I just used precision tweezers and generally ended up with slightly too big loops, which then required a lot of solder.

The switches I used also had a thicker, stiffer and a thinner, more flexible pin, although judging from your photos, the thick pin certainly wasn't as large and square as yours.  At any rate, I went with the method of preforming loops so that they can simply be placed over the pin and stay in place long enough to solder them one by one.  This should reduce wear and tear on the switch pins.  Did you stop pre-forming the loops, just to save on time/effort, or did the method prove to have any particular problems for you?

> Soldering the multi-stranded wire is really painful and was the only place where I had to re-solder quite a few, because it wasn't connected properly.

Did you tin the wires properly?  As I describe in the guide, the basic idea is to tin the wire ends, i.e. apply some flux to the wire followed by some solder, so that about 5-7mm of the end become "solid-core", as the strands are saturated with solder.  You can then similarly apply flux and solder on the target location (here the solder will simply form a coating of the target location) and finally just position the wire on the target site and reflow the solder.  With some practice, only minimal application of heat is required (although spatial constraints still make the work hard enough).

> Crimping was pretty rough.

Yes... I made sure to warn about that in the build guide.  You mention the Molex brand; did you use some other alternative for the connectors made by Molex or the parts by TE connectivity I listed in the BOM?  I'm asking because your crimps don't look properly executed, unless perhaps they're of some different type.  I can only see the insulation crimps, but these don't seem to be properly wrapped around the  insulation.  Generally this part of the crimp is for strain relief and doesn't contribute much of the total strength, so it shouldn't be the cause of your pulled out terminals, but if you used the wrong tool, or technique to execute the crimp (e.g. just pressing them with pliers), perhaps the crimps on the wire strands are similarly problematic.

> Oh, and another thing that was hard was dealing with the insulation for the solid core wire when doing the rows.

Yes and furthermore, in my case the insulation proved surprisingly incapable of preventing shorts.  I had more than one, simply because the only places where shorts can develop are close to the solder joints (where bare column leads run vertically beneath the row wire) and there it's easy not to have insulation at all, because you've removed too much of it, or to melt through the insulation while making the joint.  Perhaps it would be best to simply use bare solid core wire and use some other way of avoiding shorts...

>I haven't had much luck with that so far, because they don't fit onto my switches and I didn't have small enough file to clean that up.

The printing of the keycap stem presents several challenges: for one, getting the cross to be tight but not too tight, doesn't allow much room for error, but each material contracts to a different degree upon deposition, depending on other parameters as well, such as bed and hotend temperatures.  Geometry that works well for PETG, results in a too tight stem with ABS.  On the other hand you can't make it too loose either, because that'd require a larger stem diameter (otherwise it'd break to easily), which won't fit in the shaft of the switch.

Perhaps other geometry will work better. I considered geometry similar to the one you show, but it seems more fragile, or less stable, so I tried to avoid it, unless it proved necessary.  At any rate, I'd suggest the following before that:

1. Make sure that your slicer doesn't add support inside the stem.  Cura seems to add a single column of support at the center of the cross for whatever reason, which I need to suppress using a support blocker.

2. The `printable-keycap` function accepts more parameters than are used in the default invocations.  For instance, for the 1u DSA keycap, the default invocation reads:

```(apply printable-keycap :dsa [1 1] :convex common-dsa-convex)```

This is the same as (see the definition more more parameters, along with their default values):

```(apply printable-keycap :dsa [1 1] :convex common-dsa-convex
          :mount-cross-length 4.1 :mount-cross-width 1.19 :mount-radius 2.75)```


You might try adding these parameters and changing the default values shown above, according to your needs.  For instance, if the cross arms are not long enough to fit the keycap, increase the `:mount-cross-length` by 0.1-0.2mm or so.  If they're not wide enough, try the same for the `:mount-cross-width`.  If the stem breaks, at the cross tips, try increasing the radius a bit.  At the end of the day though, a needle file is likely to prove necessary.  Also keep a disposable switch stem handy, e.g. the stem from a stabilizer, which you can carefully force in to "break in" the keycap cross, without risking damage to an installed switch.

**Reply by hasufell** [reply]
> Did you stop pre-forming the loops, just to save on time/effort, or did the method prove to have any particular problems for you?

Yes, it was time intensive and didn't work well for the bendable pins... the pin kept breaking on my "free" switch and the loop would often not fit on the target pin.

> Did you tin the wires properly?

I guess not. I didn't play with flux much, since I had bitter problems due to flux in a previous soldering project (it caused leak current).

> did you use some other alternative for the connectors made by Molex or the parts by TE connectivity I listed in the BOM?

Yes, I used exactly the parts from the build guide.

> I'm asking because your crimps don't look properly executed, unless perhaps they're of some different type. 

That is well possible. I have a multi tool that has a crimp section, but it looks it's for much bigger ones. Any more advice here would be appreciated. I just discovered that I had some sort of short between two crimps, so I'm worried such problems will crop up again until I make it better.

Like... I'm not sure what the right crimp tool is.

**Reply by hasufell** [reply]
I tried again and I believe I got a somewhat better result:

![IMG_20210506_004147](https://user-images.githubusercontent.com/1241845/117219050-c0a3e200-ae04-11eb-8bac-8eca60907d6d.jpg)

This pin now seems to fit deeper into the header and doesn't just fall out when you pull the wire, so that's an improvement. I'm not sure if the mid-section needs crimping as well around the wire (not the insulation)?

Note: I use 22 AWG wire, although the crimp terminal says it's for 30-24. But I guess that still works. I'll try to re-crimp tomorrow. If the result looks well, maybe we could add a picture to the build guide showing a correctly crimped end.

**Comment by dpapavas** [comment]
>Yes, it was time intensive and didn't work well for the bendable pins... the pin kept breaking on my "free" switch and the loop would often not fit on the target pin.

Did you use the "free" switch pin to form the wire coil around it?  My idea was to use the switch simply to bend the wires in the right directions and mark off where to start the coil (with a kink), then to use needle-point tweezers or something similar to form the coil.  Perhaps the guide isn't clear enough on this point.

>I tried again and I believe I got a somewhat better result:

This terminal is half-crimped.  The terminal needs to be crimped at two points: the insulation crimp, which you've executed, and the wire strand crimp itself, which, as you suspect still needs to be done.  The latter (the wire crimp) is the most important part of the crimp, with the former (the insulation crimp) mostly offering strain relief.

To get an idea of what a properly executed crimp looks like (this particular terminal is what's called an "F-type open crimp terminal"), what to strive for and what to avoid, have a look at pages 13-14 [here]( https://www.fscables.com/sites/admin/plugins/elfinder/files/fscables/other%20pdf/cablecraft_crimp_guide.pdf) for instance, or in similar guides.  I suspect it is hard to get an "excellent" crimp with generic tools (I certainly couldn't get one), so you shouldn't worry too much about each and every point, but your result should be in the right ball park in every aspect.  You can also look through YouTube videos and similar tutorials, to get a more visual idea of what the process and the tools look like, but watch out, as there's plenty of bad advice in such sources.

The [NASA workmanship guides](https://workmanship.nasa.gov/lib/insp/2%20books/frameset.html) I mention in the build guide, may also be of interest, although they cover a lot of irrelevant crimp types.  Note the opening passage:

>Crimping is an efficient and highly reliable method to assemble and terminate conductors, and typically provides a stronger, more reliable termination method than that achieved by soldering.

It should be *pretty* hard to pull the wire out of a properly executed crimp.

**Reply by hasufell** [reply]
> Did you use the "free" switch pin to form the wire coil around it?

Yes exactly. The tweezers had a differing size (they get sharper towards the tip) and I could never manage a proper loop. Either it was too tight or too loose.

> This terminal is half-crimped. The terminal needs to be crimped at two points: the insulation crimp, which you've executed, and the wire strand crimp itself, which, as you suspect still needs to be done. The latter (the wire crimp) is the most important part of the crimp, with the former (the insulation crimp) mostly offering strain relief.

Excellent. I'm going to buy another crimp tool today and see if I get better results. I have about 50 crimp terminals left, so there should be something usable at the end of the day.

**Reply by hasufell** [reply]
I haven't had much luck finding a compatible crimper. The one from my local store didn't work and was too wide.

https://www.te.com/deu-de/product-2119539-1.html seems to be a compatible product, but I'm not sure I'm willing to pay 1600 dollars :roll_eyes:

**Comment by dpapavas** [comment]
The crimping tool I found at a local electronics store looks somewhat like [this](https://www.velleman.eu/products/view/?id=281906).  I did the job, but I'm reluctant to recommend it, as it didn't make crimping these small terminals very easy.  (Perhaps I just weren't using it correctly, but I did try to find the proper way.)

It crimps both sites at once, which makes it necessary to insert the terminal and wire deep into the die.  For one it's hard to hold the terminal (I had to thread it into an unwrapped paperclip) and for another, it's hard to see what you're doing and if you've inserted it far enough, or too far.  Getting the strands fully crimped, without having them extend too far past the crimp and without getting the insulation in there with them can take some practice and even then, I couldn't get a completely satisfactory result (although for the trivially low currents these wires have to carry, it shouldn't be an issue).  Note that my tool seemed to be a relatively cheap copy of something like this, so perhaps a better made tool will perform better.  (For instance, my tool had a "step" in the die, where the open crimping tabs of the terminal butted up against so it was easy to insert the terminal itself correctly, but from then on, there was no guide for inserting the wire properly.  Some tools seem to have adjustable stops where the insulation of the wire butts up against, which might make proper crimping easier.)

There are other tools, such as [this](https://grobotronics.com/crimping-pliers-for-jst-engineer-pa-09.html), which take a longer time to fully crimp a terminal, as you have to separately crimp each site, but seem to be easier to handle, as you can see exactly where the wire is inside the terminal and where you're crimping.
